rfd/0209-mcp-access.md

```diff
@@ -0,0 +1,359 @@
+Teleport logs.
+
+Incoming bytes from the client do get the following treatments before they are
+forwarded to the launched MCP server:
+- Parsed into MCP/JSON-RPC messages. 
+- If the message is a `tools/call` request, use the access-checker to confirm
+  whether the identity has access to the requested tool.
+  - If the `tools/call` is access denied, make a text result detailing this
+    error and send it back to the client connection. Do not forward the request to
+    the launched MCP server in this case.
+- Record audit events for eligible messages.
+
+The launched MCP server should be stopped once the connection ends.
+
+#### Implementation - "tsh mcp" commands
+
+A suite of `tsh mcp` commands are added on the client side:
+- `tsh mcp ls`: lists MCP server apps, with slightly different columns from `tsh
+  app ls`.
+- `tsh mcp connect`: command to be called by the AI tool to launch an MCP server
+  through Teleport. User cert with `route_to_app` is automatically obtained if
+  the app hasn't been logged in yet. stdin and stdout of this process are
+  proxied to Teleport backend. Any local tsh logs are sent to stderr.
+- `tsh mcp config`: shows a sample JSON configuration that launches the MCP
+  server with `tsh mcp connect` for Claude Desktop. (Most AI tools use the same
+  JSON format as well.)
+- `tsh mcp db`: command to be called by the AI tool to launch a local MCP server
+  that provides a database query tool through Teleport Database access. This
+  feature is outside this RFD but shares the same `tsh mcp` root.
+
+#### Compatibility with App access
+
+Since the MCP servers run as apps, here’s a list of some existing app
+functionalities and how they align with it:
+- Web UI: MCP servers will be listed as Application resources. Similar to TCP
+  apps, MCP servers cannot be "launched" through Web UI, and clicking "Connect"
+  button opens a dialog with `tsh` instructions.
+- VNet: not supported.
+- Per-session MFA: not supported.
+- Access Request: supported.
+- Dynamic registration: supported.
+- Audit events: will generate `cert.create` event when cert is requested, but no
+  `app.session.start` event.
+- `tsh`:
+  - `tsh app ls`: will be listed with `MCP` as the type.
+  - `tsh app login`: supported, but not necessary.
+  - `tsh app logout`: supported, but why.
+  - `tsh app config`: not supported, redirect user to `tsh mcp config`.
+  - `tsh proxy app`: not supported.
+
+#### Usage reporting
+
+`mcp.session.start` events will be reported for existing session start events
+(`tp.session.start.v2`), with `mcp` as the session type. MCP sessions will be
+counted as regular `app_sessions` for user activity temporarily until new
+`mcp_sessions` is added on the cloud side.
+
+### Security
+
+Transport and auth wise, this feature is mostly the same with existing TCP
+access with the addition of the new role option `mcp_tools`.
+
+One potential concern is that the App service executes an arbitrary 3rd party
```

Comment:
I had exactly the same thought - should we consider using fork/reexec approach to execute these similar to ssh?

Reply:
Thinking about this some more, fork/reexec with cgroups probably won't really do much on its own without proper namespaces support which is a bit of work. I'm thinking providing (and maybe requiring) and option of downgrading the uid/gid and recommending to use docker to run these is probably a good start.

We could even go as far as requiring to run these in docker. Most of [MCP server examples](https://github.com/modelcontextprotocol/servers/tree/main/src) have docker images anyway. This would affect the UX though.

Reply:
@Tener I think it's a great idea, I would imagine docker and k8s runtimes would be very popular from the start.

Reply:
Added mandatory field `run_as_local_user` with examples in the UX section, including an example running docker command.

Also added `streamable HTTP transport` support for phase 2. Then we can always recommend that over stdio-based.